### PR TITLE
feat: determine if the host has a gpu

### PR DESCRIPTION
**Pull request by srx0914**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -56,22 +57,28 @@ def Stream():
...
-                if result["bboxes2draw"]:
-                    bboxes.append(result["bboxes2draw"])
-            else:
-                if bboxes:
-                    frame = plot_bboxes(im, bboxes[0])
+    if not torch.cuda.is_available():
+        while True:
+            if deq:
+                im = deq.popleft()
+                pipe.stdin.write(im.tobytes())
```
Comment (ghost) [inline main.py:69]:
```python
if deq:
    ...
else:
    time.sleep(1)
```

File: main.py [hunk 2]
```diff
@@ -56,22 +57,28 @@ def Stream():
...
+    else:
```
Comment (ghost) [inline main.py]:
```python
if not torch.cuda.is_available():
    ...
else:  # 去掉
```

File: main.py [hunk 3]
```diff
@@ -56,22 +57,28 @@ def Stream():
...
+        while True:
+            if deq:
```
Comment (ghost) [inline main.py]: if deq 后面应该有对应的 else，然后 sleep(1)

**Review by ghost — CHANGES_REQUESTED** [review]
```python
def Receive():
    print("start Reveive ...")
    while True:
        cap = cv2.VideoCapture(os.getenv("rtsp"))
        cap.set(cv2.CAP_PROP_BUFFERSIZE, 2)
        while cap.isOpened():
            _, frame = cap.read()
            deq.append(frame)
        print('deq is empty, retry')
        time.sleep(1)
```

File: main.py [hunk 4]
```diff
@@ -13,11 +15,16 @@
...
-        deq.append(frame)
+    while True:
+        cap = cv2.VideoCapture(os.getenv("rtsp"))
+        cap.set(cv2.CAP_PROP_BUFFERSIZE, 2)
+        while cap.isOpened():
+            _, frame = cap.read()
+            deq.append(frame)
+        timeArr = time.localtime(int(time.time()))
+        other_StyleTime = time.strftime("%Y-%m-%d %H:%M:%S", timeArr)
+        print("deq is empty, retry, time:{}".format(other_StyleTime))
```
Comment (ghost) [inline main.py]: 太复杂了，这样吧

```python
print("[{} UTC] deq is empty, retry...".format(datetime.datetime.utcnow()))
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: main.py [hunk 5]
```diff
@@ -13,11 +16,14 @@
...
-    while cap.isOpened():
-        _, frame = cap.read()
-        deq.append(frame)
+    while True:
+        cap = cv2.VideoCapture(os.getenv("rtsp"))
+        cap.set(cv2.CAP_PROP_BUFFERSIZE, 2)
+        while cap.isOpened():
+            _, frame = cap.read()
+            deq.append(frame)
+        print("[{} UTC] deq is empty, retry...".format(datetime.datetime.utcnow()))
```
Comment (ghost) [inline main.py]: print("[{} UTC] rtsp stream is closed, retry...".format(datetime.datetime.utcnow()))

File: main.py [hunk 6]
```diff
@@ -56,6 +62,13 @@ def Stream():
     i = 1
     count = 5
     bboxes = []
+    if not torch.cuda.is_available():
+        while True:
+            if deq:
+                im = deq.popleft()
+                pipe.stdin.write(im.tobytes())
+            else:
+                time.sleep(1)
```
Comment (ghost) [inline main.py:71]: print("[{} UTC] deq is empty, retry...".format(datetime.datetime.utcnow()))

File: main.py [hunk 7]
```diff
@@ -72,6 +85,8 @@ def Stream():
                     frame = im
             pipe.stdin.write(frame.tobytes())
             i = (i + 1) % count
+        else:
+            time.sleep(1)
```
Comment (ghost) [inline main.py:90]: print("[{} UTC] deq is empty, retry...".format(datetime.datetime.utcnow()))

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by FrankGTY — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**